src/opm/input/eclipse/Schedule/KeywordHandlers.cpp

```diff
@@ -1245,23 +1245,39 @@ File {} line {}.)", wname, location.keyword, location.filename, location.lineno)
+            if (conn_defaulted(record)) {
+                const auto new_well_status = Well::StatusFromString(status_str);
+
                 for (const auto& wname : well_names) {
-                    {
-                        const auto& well = this->getWell(wname, currentStep);
-                        if( well_status == open && !well.canOpen() ) {
-                            auto elapsed = this->snapshots[currentStep].start_time() - this->snapshots[0].start_time();
-                            auto days = std::chrono::duration_cast<std::chrono::hours>(elapsed).count() / 24;
-                            std::string msg = "Well " + wname
-                                + " where crossflow is banned has zero total rate."
-                                + " This well is prevented from opening at "
-                                + std::to_string( days ) + " days";
-                            OpmLog::note(msg);
-                        } else {
-                            this->updateWellStatus( wname, currentStep, well_status);
-                            if (handlerContext.sim_update)
-                                handlerContext.sim_update->affected_wells.insert(wname);
+                    if ((new_well_status == open) && !this->getWell(wname, currentStep).canOpen()) {
+                        auto elapsed = this->snapshots[currentStep].start_time() - this->snapshots[0].start_time();
+                        auto days = std::chrono::duration_cast<std::chrono::hours>(elapsed).count() / 24;
+                        std::string msg = "Well " + wname
+                            + " where crossflow is banned has zero total rate."
+                            + " This well is prevented from opening at "
+                            + std::to_string( days ) + " days";
+                        OpmLog::note(msg);
+                    }
+                    else {
+                        const auto did_update_well_status =
+                            this->updateWellStatus(wname, currentStep, new_well_status);
+
+                        if (handlerContext.sim_update) {
+                            handlerContext.sim_update->affected_wells.insert(wname);
+                        }
+
+                        if (did_update_well_status && (new_well_status == open)) {
+                            // Record possible well injection/production status change
+                            auto well2 = this->snapshots[currentStep].wells.get(wname);
+
+                            const auto did_flow_update =
+                                (well2.isProducer() && well2.updateHasProduced())
+                                ||
+                                (well2.isInjector() && well2.updateHasInjected());
+
+                            if (did_flow_update) {
+                                this->snapshots[currentStep].wells.update(std::move(well2));
+                            }
```

Comment:
I don't see why this copy-move construct is needed and I think it might be even dangerous because an internal shared_ptr is reassigned which might be unexpected elsewhere.

Why not simply:
```
auto& well2 = this->snapshots[currentStep].wells.get(wname);
if(well2.isProducer()) well2.updateHasProduced();
if(well2.isInjector()) well2.updateHasInjected();
``
There is a get method returning a reference and it seems easier to read.

Reply:
> I don't see why this copy-move construct is needed and I think it might be even dangerous

This is the expected pattern and it is used throughout the this part of the code.  Note that despite appearances, `currentStep` is really the latest report step so we're changing the last element of `snapshots`.

Reply:
Ok, not dangerous as long as everybody behaves as expected, still inefficient and unclear. Maybe improve the expected?

Reply:
>>> I don't see why this copy-move construct is needed and I think it might be even dangerous
>>
>> This is the expected pattern and it is used throughout this part of the code. Note that despite appearances, `currentStep` is really the latest report step so we're changing the last element of snapshots.
>
> Ok, not dangerous as long as everybody behaves as expected, still inefficient and unclear. Maybe improve the expected?

Actually I take back my earlier statement.  It is in fact generally dangerous to **not** take the copy here.  In this particular spot we can get by using a raw reference and calling the appropriate `update*()` member function if we choose to explicitly embed knowledge that the [earlier call to `updateWellStatus()`](https://github.com/OPM/opm-common/blob/4150936e7c0f52fccbf214dbbe3266083de0fef3/src/opm/input/eclipse/Schedule/KeywordHandlers.cpp#L1262-L1263) created a new well object.  In general, however, calling any function named `update*()` directly on the object referenced by `snapshots[i].wells.get(name)` will potentially alter the state of well objects earlier in the schedule and create bad inconsistencies.

Reply:
Thanks for the explanation. Now I get it. Seems rather fragile (what is possible will sooner or later be tried...). I guess this will need to be enforced by the code eventually, then.

Reply:
>  Seems rather fragile

I don't disagree.  The restrictions come from having containers of `shared_ptr<>` in an effort to reduce memory consumption.  That way, multiple report steps can share the same underlying object if its controls don't change across those report steps.